### ISSUE TITLE
Useful Papers

### ISSUE THREAD

**Issue by moefarhat** [chriswernette/eecs504-final-project#2]
### This paper uses neural nets to recognize company logos on billboards, uses flikr as a source of learning.
- https://www.cc.gatech.edu/~hays/7476/projects/Punarva_Prabhudev.pdf

**Comment by amcrean** [comment]
https://www.semanticscholar.org/paper/Real-Time-Billboard-Substitution-in-a-Video-Stream-Medioni-Guy/b86c46dd39239407704efc7bf78ffde68b0fd094

**Comment by pmpaquet** [comment]
http://delivery.acm.org/10.1145/1110000/1101221/p343-li.pdf?ip=35.3.89.14&id=1101221&acc=ACTIVE%20SERVICE&key=93447E3B54F7D979%2E0A17827594E6F2C8%2E4D4702B0C3E38B35%2E4D4702B0C3E38B35&__acm__=1541800408_8928cd907e75170e1f053390a52fec11

https://ieeexplore.ieee.org/abstract/document/4160282

**Comment by chriswernette** [comment]
@moefarhat - I found your original paper you linked was more from a comparing results standpoint and not as much how to implement. But they had great links off that paper to others. 
**In order of usefulness.**
[DeepLogo: Hitting Logo Recognition with the Deep Neural Network Hammer](https://arxiv.org/pdf/1510.02131.pdf)
[Scalable Triangulation-based Logo Recognition](https://www.researchgate.net/publication/221318676_Scalable_Triangulation-based_Logo_Recognition)
[ImageNet Classification with Deep Convolutional Neural Networks](https://www.nvidia.cn/content/tesla/pdf/machine-learning/imagenet-classification-with-deep-convolutional-nn.pdf)

**I also found a few other papers, which might be worthwhile.**
[Scalable Logo Detection and Recognition with Minimal Labeling](https://ieeexplore.ieee.org/abstract/document/8396990)
[Logo retrieval with a contrario visual query expansion](https://dl.acm.org/citation.cfm?id=1631361)
[Scalable logo recognition in real-world images](https://dl.acm.org/citation.cfm?id=1992021)
[SSD](https://towardsdatascience.com/logo-detection-in-images-using-ssd-bcd3732e1776)
[Deep Learning for Logo Detection](https://shuaiw.github.io/2017/01/24/deep-learning-for-logo-recognition-part-i.html)
[Logo Detection Using Machine Learning](https://www.researchgate.net/post/Logo_Detection_from_Image_Using_machine_learning)
[Research Gate Logo Detection General Answer](https://www.researchgate.net/post/Logo_Detection_from_Image_Using_machine_learning)

**Github projects:**
[Logo Detection - YOLOv2](https://github.com/akarshzingade/Logo-Detection-YOLOv2)
[YOLOv2 Logo Detection Medium Article](https://medium.com/@akarshzingade/logo-detection-using-yolov2-8cda5a68740e)
[Logo Detection](https://github.com/AchmnShukla/Logo-Detection)
[Logo Detection Deep Learning](https://flovv.github.io/Logo_detection_deep_learning/)

**Comment by chriswernette** [comment]
**For the billboard detection (in no particular order of usefulness):**

**Hough Transform/Edge Detection Methods**
I think this is cheap, easy to implement and will work relatively well. However it won't be state of the art or most accurate. I think we should probably try to implement one of these methods.

[Stack Overflow Billboard Corner Detection](https://stackoverflow.com/questions/52814400/billboard-corner-detection)
[How to detect parallelograms from the detected edge points in python](https://stackoverflow.com/questions/47015716/how-to-detect-parallelograms-from-the-detected-edge-points-in-python)
[Parallelogram Detection Using the Tiled Hough Transform](http://inf.ufrgs.br/~crjung/papers/paper_1125.pdf)
[OpenCV shape detection](https://www.pyimagesearch.com/2016/02/08/opencv-shape-detection/)
[Stack Overflow find intersection point of two lines drawn using houghlines opencv](https://stackoverflow.com/questions/46565975/find-intersection-point-of-two-lines-drawn-using-houghlines-opencv)
[Billboard Detection - Ad Agency Real Implementation, just high level](http://tech.adstruc.com/post/28306434540/billboard-detection)

Read section 4 of this paper, should be very helpful for detecting edges of billboards. He uses something he calls _Smartlines_. In this paper they detect grass and remove it from the image they're searching - I really like the idea of detecting sky or clouds and removing them from the search domain.
[Automatic Billboard Detection Master's Thesis (more for soccer billboards)](http://citeseerx.ist.psu.edu/viewdoc/download;jsessionid=96199FCCBE3AB0687D6589E1AF52ECD7?doi=10.1.1.392.894&rep=rep1&type=pdf)

**Objectness**
I haven't looked enough into this method, but I assume it is also a ML type method that might be very expensive.
[BING: Binarized Normed Gradients for Objectness Estimation at 300fps](https://www.robots.ox.ac.uk/~vgg/rg/papers/14cvprObjectnessBING.pdf)
[Measuring the objectness of image windows](http://calvin.inf.ed.ac.uk/wp-content/uploads/Publications/alexe12pami.pdf)
[Objectness measure V2.2](http://groups.inf.ed.ac.uk/calvin/objectness/)

**Contour Detection/Instance Segmentation**
This seems to be the state of the art method, with Contour Detection getting us segmentation of the billboards, and then we'd have to come up with an algorithm that says a parallelogram shape is most likely a billboard. Instance Segmentation would train directly from billboard images and then use a CNN to get us an exact image mask of the billboard. I am concerned about the expense involved to train/run this type of algorithm. Might need GPUs.

[Discriminative Generative Contour Detection](http://faculty.ucmerced.edu/mhyang/papers/bmvc13_contour.pdf)
[Contour Detection, Image and Video Segmentation](https://www2.eecs.berkeley.edu/Research/Projects/CS/vision/grouping/index2.html)

Read section 6. Uses instance segmentation to solve the problem, i.e. a deep learning/CNN approach where you train a network to recognize billboards and then it can segment any image automatically as a billboard or not. This is a really good example of how to do it via ML instead of basic computer vision techniques.
[Suppressing distractors from driver's field of view - Master's Thesis](https://dspace.cvut.cz/bitstream/handle/10467/77075/F3-DP-2018-Kolarik-Tomas-kolarto4_DP.pdf)

I think this and that master's thesis are the most useful if we want to go the ML route. It seems like instance segmentation is exactly what we need, but contour detection with some model for what a billboard constitutes could also get us there.
[Splash of Color: Instance Segmentation with Mask R-CNN and TensorFlow](https://engineering.matterport.com/splash-of-color-instance-segmentation-with-mask-r-cnn-and-tensorflow-7c761e238b46)

Great tutorial on how to use R-mask CNN for instance segmentation. He makes it sound not so bad.
[Deep learning based Object Detection and Instance Segmentation using Mask R-CNN in OpenCV (Python / C++)](https://www.learnopencv.com/deep-learning-based-object-detection-and-instance-segmentation-using-mask-r-cnn-in-opencv-python-c/)

Other:
I would consider this one cheating because they have manually annotated billboards in their soccer images and then just just build up a probability PDF of where the billboard is in the image using different descriptors. Because they have manually annotated billboards, they know what is part of the billboard and what isn't, so they can easily make a mask of what to project and what to leave out. Some of the interesting parts are the motion tracking maybe? Seems like a lot of 568 concepts are used.
[Visual Tracking and Localisation of Billboards in Streamed Soccer Matches](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.9.7148&rep=rep1&type=pdf)

**Comment by chriswernette** [comment]
[How to use the Hough Transform for Line/Lane Detection](https://www.youtube.com/watch?v=KEYzUP7-kkU)
[Link to source code and files](https://pysource.com/2018/03/07/lines-detection-with-hough-transform-opencv-3-4-with-python-3-tutorial-21/)

**Comment by chriswernette** [comment]
**Practical Resources for the Project**
[Canny in OpenCV](https://docs.opencv.org/3.1.0/da/d22/tutorial_py_canny.html)
[Loading in, manipulating images](https://docs.opencv.org/3.0-beta/doc/py_tutorials/py_gui/py_image_display/py_image_display.html)
[Geometric Transforms of Images in OpenCV - rotation, affine, homography, etc](https://docs.opencv.org/3.4/da/d6e/tutorial_py_geometric_transformations.html)
[Same as above but different flavor](https://docs.opencv.org/2.4/modules/imgproc/doc/geometric_transformations.html)
[OpenCV Misc Color Transformations - eg. BGR2GRAY, BGR2RGB](https://docs.opencv.org/3.1.0/d7/d1b/group__imgproc__misc.html#ga397ae87e1288a81d2363b61574eb8cab)
[Find Intersection of Two Lines](https://en.wikipedia.org/wiki/Line%E2%80%93line_intersection#Given_two_points_on_each_line)
[Intersection of 2 lines code](https://www.geeksforgeeks.org/program-for-point-of-intersection-of-two-lines/)
[Intersection of 2 lines BEAST MODE CODE](https://stackoverflow.com/questions/46565975/find-intersection-point-of-two-lines-drawn-using-houghlines-opencv)

Need to research this more, but the find contours function was looking almost identical to the Canny Edges for me :(
[OpenCV Find Contours - could maybe replace Hough Lines?](https://docs.opencv.org/3.4/d4/d73/tutorial_py_contours_begin.html)
[OpenCV Convex Hull](https://www.learnopencv.com/convex-hull-using-opencv-in-python-and-c/)
[Process Image to find external contour](https://stackoverflow.com/questions/25733694/process-image-to-find-external-contour)
[Finding extreme points in contours using openCV](https://www.pyimagesearch.com/2016/04/11/finding-extreme-points-in-contours-with-opencv/)
[Scipy Convex Hull Implementation - Maybe better?](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.ConvexHull.html)
[Convex Hull by hand using only numpy](https://www.oreilly.com/ideas/an-elegant-solution-to-the-convex-hull-problem)

Linear Regression/Least Squares Fitting a Rectangle:
[Fitting a rectangle](https://www.inf.ethz.ch/personal/arbenz/MatlabKurs/node88.html)
[Best fitted rectangle on a set of points](https://math.stackexchange.com/questions/2711509/best-fitted-rectangle-on-a-set-of-points)
[Fit a rectangle around points](https://stackoverflow.com/questions/34479435/fit-rectangle-around-points)

How to convert videos to images:
[OpenCV - Getting Started with Videos](https://docs.opencv.org/3.0-beta/doc/py_tutorials/py_gui/py_video_display/py_video_display.html)
[Learn OpenCV - Read, Write and Display a video using OpenCV](https://www.learnopencv.com/read-write-and-display-a-video-using-opencv-cpp-python/)
[Scikit-image working with videos](http://scikit-image.org/docs/dev/user_guide/video.html)

RDP Algorithm (to simplify polygons, not sure if useful):
[RDP on Wikipedia](https://en.wikipedia.org/wiki/Ramer%E2%80%93Douglas%E2%80%93Peucker_algorithm)
[RDP Overview](https://ianqvist.blogspot.com/2010/05/ramer-douglas-peucker-polygon.html)
[RDP Library #1](https://pypi.org/project/rdp/)
[RDP Library #2](https://pypi.org/project/simplification/)
